File: shared/ui/ipc/webview.protocol.ts
import { CodemarkType } from "@codestream/protocols/api";
import { NotificationType } from "vscode-jsonrpc";
import { Range } from "vscode-languageserver-types";
import { IpcRoutes } from "./webview.protocol.common";

export * from "./webview.protocol.common";

export * from "./host.protocol";
export * from "./host.protocol.editor";
export * from "./host.protocol.notifications";
export * from "./host.protocol.review";
export * from "./host.protocol.vsls";

export * from "./webview.protocol.notifications";

// TODO: This should be a request to the webview -- not a notification
export interface ShowCodemarkNotification {
	codemarkId: string;
	sourceUri?: string;
}
export const ShowCodemarkNotificationType = new NotificationType<ShowCodemarkNotification, void>(
	`${IpcRoutes.Webview}/codemark/show`
);

// TODO: This should be a request to the webview -- not a notification
export interface ShowReviewNotification {
	reviewId: string;
	sourceUri?: string;
}
export const ShowReviewNotificationType = new NotificationType<ShowReviewNotification, void>(
	`${IpcRoutes.Webview}/review/show`
);

// TODO: This should be a request to the webview -- not a notification
export interface ShowPullRequestwNotification {
	providerId: string;
	id: string;
	commentId?: string;
	/* Either a providerId && id are used, OR a url */
	url?: string;
	source?: string;
}
export const ShowPullRequestNotificationType = new NotificationType<
	ShowPullRequestwNotification,
	void
>(`${IpcRoutes.Webview}/pullRequest/show`);

// TODO: This should be a request to the webview -- not a notification
export interface ShowStreamNotification {
	streamId: string;
	threadId?: string;
	codemarkId?: string;
}
export const ShowStreamNotificationType = new NotificationType<ShowStreamNotification, void>(
	`${IpcRoutes.Webview}/stream/show`
);

// TODO: This should be a request to the webview -- not a notification
export interface NewCodemarkNotification {
	uri?: string;
	range?: Range;
	type: CodemarkType;
	source?: string;
}

export const NewCodemarkNotificationType = new NotificationType<NewCodemarkNotification, void>(
	`${IpcRoutes.Webview}/codemark/new`
);

export interface NewReviewNotification {
	uri?: string;
	range?: Range;
	source?: string;
	includeLatestCommit?: boolean;
}

export interface NewPullRequestBranch {
	name: string;
	remote?: {
		name: string;
		provider?: {
			id: string;
			name: string;
			domain: string;
		};
		url?: string;
	};
	repoPath: string;
}

export interface NewPullRequestNotification {
	uri?: string;
	range?: Range;
	source?: string;
	branch?: NewPullRequestBranch;
}

export const NewReviewNotificationType = new NotificationType<NewReviewNotification, void>(
	`${IpcRoutes.Webview}/review/new`
);

export const NewPullRequestNotificationType = new NotificationType<
	NewPullRequestNotification,
	void
>(`${IpcRoutes.Webview}/pullRequest/new`);

export interface StartWorkNotification {
	uri?: string;
	source?: string;
}

export const StartWorkNotificationType = new NotificationType<StartWorkNotification, void>(
	`${IpcRoutes.Webview}/work/start`
);

export interface ShowNextChangedFileNotification {}

export const ShowNextChangedFileNotificationType = new NotificationType<
	ShowNextChangedFileNotification,
	void
>(`${IpcRoutes.Webview}/showChangedFile/next`);

export interface ShowPreviousChangedFileNotification {}

export const ShowPreviousChangedFileNotificationType = new NotificationType<
	ShowPreviousChangedFileNotification,
	void
>(`${IpcRoutes.Webview}/showChangedFile/previous`);

File: vs/src/CodeStream.VisualStudio.Core/Models/Webview.Protocol.cs
using System;

namespace CodeStream.VisualStudio.Core.Models {
	public class ShowCodemarkNotification {
		public string CodemarkId { get; set; }
		public string SourceUri { get; set; }
	}

	public class ShowCodemarkNotificationType : NotificationType<ShowCodemarkNotification> {
		public const string MethodName = "webview/codemark/show";
		public override string Method => MethodName;
	}

	public class ShowStreamNotification {
		public string StreamId { get; set; }
		public string ThreadId { get; set; }
		public string CodemarkId { get; set; }
	}

	public class ShowStreamNotificationType : NotificationType<ShowStreamNotification> {
		public const string MethodName = "webview/stream/show";
		public override string Method => MethodName;
	}

	public class NewCodemarkNotification {
		public NewCodemarkNotification(Uri uri, Range range, CodemarkType type, string source) {
			Uri = uri.ToString();
			Range = range;
			Type = type;
			Source = source;
		}

		public string Uri { get; }
		public Range Range { get; }
		public CodemarkType Type { get; }
		public string Source { get; }
	}

	public class NewCodemarkNotificationType : NotificationType<NewCodemarkNotification> {
		public const string MethodName = "webview/codemark/new";
		public override string Method => MethodName;
	}

	public class StartWorkNotification {
		public StartWorkNotification(string source, Uri uri = null) {
			Source = source;
			Uri = uri?.ToString();
		}

		public string Source { get; }
		public string Uri { get; }
	}

	public class StartWorkNotificationType : NotificationType<StartWorkNotification> {
		public const string MethodName = "webview/work/start";
		public override string Method => MethodName;
	}

	public class NewReviewNotification {
		public NewReviewNotification(Uri uri, string source) {
			Uri = uri?.ToString();
			Source = source;
		}
		public string Uri { get; }
		public string Source { get; }
	}

	public class NewReviewNotificationType : NotificationType<NewReviewNotification> {

		public const string MethodName = "webview/review/new";
		public override string Method => MethodName;
	}

	public class ShowNextChangedFileNotification { }
	public class ShowNextChangedFileNotificationType : NotificationType<ShowNextChangedFileNotification> {

		public const string MethodName = "webview/showChangedFile/next";
		public override string Method => MethodName;
	}

	public class ShowPreviousChangedFileNotification { }
	public class ShowPreviousChangedFileNotificationType : NotificationType<ShowPreviousChangedFileNotification> {

		public const string MethodName = "webview/showChangedFile/previous";
		public override string Method => MethodName;
	}
}
